File: examples/GExternalEvaluator/evaluator/evaluator.cpp
/**
 * @file evaluator.cpp
 */

/* Copyright (C) Dr. Ruediger Berlich and Karlsruhe Institute of Technology
 * (University of the State of Baden-Wuerttemberg and National Laboratory
 * of the Helmholtz Association)
 *
 * Contact: info [at] gemfony (dot) com
 *
 * This file is part of the Geneva library, Gemfony scientific's optimization
 * library.
 *
 * Geneva is free software: you can redistribute it and/or modify
 * it under the terms of version 3 of the GNU Affero General Public License
 * as published by the Free Software Foundation.
 *
 * Geneva is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with the Geneva library.  If not, see <http://www.gnu.org/licenses/>.
 *
 * For further information on Gemfony scientific and Geneva, visit
 * http://www.gemfony.com .
 */

/*
 * This program performs a value calculation for parameters that have been
 * handed to it by the Geneva library. It serves as an example on how it is
 * possible to use external evaluation programs with the Geneva library.
 */

// standard headers go here
#include <iostream>
#include <cstdlib>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <cmath>

// Boost headers go here
#include <boost/random.hpp>
#include <boost/date_time.hpp>

// Geneva headers go here
#include "commandLineParser.hpp"
#include "GDataExchange.hpp"

const std::size_t PARABOLADIM=1000;
const double PARABOLAMIN=-100.;
const double PARABOLAMAX=100.;

int main(int argc, char **argv)
{
	Gem::Util::GDataExchange ge;
	boost::uint16_t executionMode, transferMode;
	bool binary=true;
	std::string paramfile;
	std::string identifyer;

	// Parse the command line
	if(!parseCommandLine(argc, argv,
												executionMode,
												paramfile,
												transferMode,
												identifyer))
	{exit(1);}

	switch(transferMode) {
	case 0:
		binary=true;
		break;
	case 1:
		binary=false;
		break;
	default:
		{
			std::cerr << "Error: Invalid transfer mode" << transferMode << std::endl;
			exit(1);
		}
	}

	// See the file commandLineParser.cpp for the available modes
	switch(executionMode) {
	case 1:	// Perform initialization code
		std::cout << "Initializing ...";
		// put your code here
		std::cout << "... done." << std::endl;
		return 0;
		break;

	case 2:	// Perform finalization code
		std::cout << "Finalizing ...";
		// put your code here
		std::cout << " ... done." << std::endl;
		return 0;
		break;

	case 3: // Evaluate
		{
			// Read in the parameter data
			ge.readFromFile(paramfile, binary);

			// Loop over all parameter sets and do the actual calculation
			int i=0;
			do {
				double result = 0.;
				// Loop over all double values and calculate summed squares ("a parabola")
				std::size_t nDoubles = ge.size<double>(); // The amount of doubles in the current data set
				for(std::size_t pos = 0; pos<nDoubles; pos++) result += pow(ge.at<double>(pos), 2);
				ge.setValue(result);
			}
			while(ge.nextDataSet());

			// Write out the results. We only want to write out one (the best) item
			ge.writeToFile(paramfile, binary, 1, true);
		}
		break;

	case 4: // Write out template
		{
			// Here we simply want PARABOLADIM double values with boundaries [PARABOLAMIN:PARABOLAMAX]
			for(std::size_t i=0; i<PARABOLADIM; i++) {
				ge.append(boost::shared_ptr<Gem::Util::GDoubleParameter>(new Gem::Util::GDoubleParameter(100., PARABOLAMIN, PARABOLAMAX)));
			}

			ge.writeToFile(paramfile, binary);
		}
		break;

	case 5: // Write out template, initializing with random variables
		{
			boost::posix_time::ptime t1;
		    boost::uint32_t seed = (uint32_t)t1.time_of_day().total_milliseconds();
			boost::lagged_fibonacci607 lf(seed);

			for(std::size_t i=0; i<PARABOLADIM; i++) {
				ge.append(boost::shared_ptr<Gem::Util::GDoubleParameter>(new Gem::Util::GDoubleParameter(PARABOLAMIN+lf()*(PARABOLAMAX-PARABOLAMIN), PARABOLAMIN, PARABOLAMAX)));
			}

			ge.writeToFile(paramfile, binary);
		}
		break;

	case 6: // Write out the result for a given parameter set
		{
			// Output the identifyer
			if(identifyer != DEFAULTIDENTIFYER) {
				std::cout << "Printing result with identifyer = " << identifyer << std::endl;
			}

			// Read in the parameter data
			// ge.readFromFile(paramfile, binary);

			// Output the data on the console
			// std::size_t nDoubles = ge.size<double>(); // The amount of doubles in the current data set
			// for(std::size_t pos = 0; pos<nDoubles; pos++) std::cout << ge.at<double>(pos) << std::endl;
		}
		break;

	default:
		std::cerr << "Error: Found invalid execution mode" << std::endl;
		exit(1);
	};


	return 0; // make the compiler happy
}

File: examples/geneva/12_GExternalEvaluator/scripts/evaluator.py
#!/usr/bin/python

################################################################################
#
# Copyright (C) Gemfony scientific UG (haftungsbeschraenkt)
#
# See the AUTHORS file in the top-level directory for a list of authors.
#
# Contact: contact [at] gemfony (dot) eu
#
# This file is part of the Geneva library collection.
#
# Geneva was developed with kind support from Karlsruhe Institute of
# Technology (KIT) and Steinbuch Centre for Computing (SCC). Further
# information about KIT and SCC can be found at http://www.kit.edu/english
# and http://scc.kit.edu .
#
# Geneva is free software: you can redistribute and/or modify it under
# the terms of version 3 of the GNU Affero General Public License
# as published by the Free Software Foundation.
#
# Geneva is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with the Geneva library. If not, see <http://www.gnu.org/licenses/>.
#
# For further information on Gemfony scientific and Geneva, visit
# http://www.gemfony.eu .
#
################################################################################

import argparse
import os
import sys
import xml.etree.ElementTree as ET

from datetime import datetime
from textwrap import dedent


def init():
    """
    Perform initialization work, for instance to set up the optimization environment.
    """
    print("Initializing the optimization environment...")
    # Nothing to do...

def finalize():
    """
    Perform finalization work, for instance to clean up the optimization environment.
    """
    print("Cleaning up the optimization environment...")
    # Nothing to do...

def setup(setup_file, initial_values):
    """
    Provide the problem setup data inside a file of the given name.
    """

    # We write the XML "by-hand", as ElementTree does not support "pretty-printing"
    header = """\
    <!--
        Problem setup file created by """ + command_name() + """
        on """ + current_time() + """
    -->
    <batch>
      <dataType>setup_data</dataType>
      <runID>0</runID>
      <nIndividuals>1</nIndividuals>
      <individuals>
        <individual0>
          <type>GParameterSet</type>
          <nVars>4</nVars>
          <vars>\n"""

    content = header
    for i in range(4):
        content += "            <var" + str(i) + ">\n"
        content += "              <name>coord_" + str(i) + "</name>\n"
        content += "              <type>GConstrainedDoubleObject</type>\n"
        content += "              <baseType>double</baseType>\n"
        content += "              <isLeaf>true</isLeaf>\n"
        content += "              <nVals>1</nVals>\n"
        if initial_values == "min":
            content += "              <values>\n"
            content += "                <value0>-10.0</value0>\n"
            content += "              </values>\n"
            content += "              <lowerBoundary>-10.0</lowerBoundary>\n"
            content += "              <upperBoundary>10.0</upperBoundary>\n"
            content += "              <initRandom>false</initRandom>\n"
        elif initial_values == "max":
            content += "              <values>\n"
            content += "                <value0>10.0</value0>\n"
            content += "              </values>\n"
            content += "              <lowerBoundary>-10.0</lowerBoundary>\n"
            content += "              <upperBoundary>10.0</upperBoundary>\n"
            content += "              <initRandom>false</initRandom>\n"
        else:
            content += "              <values>\n"
            content += "                <value0>0.0</value0>\n"
            content += "              </values>\n"
            content += "              <lowerBoundary>-10.0</lowerBoundary>\n"
            content += "              <upperBoundary>10.0</upperBoundary>\n"
            content += "              <initRandom>true</initRandom>\n"
        content += "          </var" + str(i) + ">\n"

    content += """\
          </vars>
          <nBounds>0</nBounds>
          <nResults>1</nResults>
        </individual0>
      </individuals>
    </batch>
    """

    with open(setup_file, 'w') as infile:
        infile.write(dedent(content))

def evaluate(in_file, out_file):
    """
    Evaluate the result for the given input parameters, and
    generate an output file.
    """
    (iteration, it_id, result) = process_input(in_file)
    write_output(out_file, iteration, it_id, result)

def archive(in_file):
    """
    Evaluate the result for the given input parameters, and
    archive the results.
    """
    (iteration, it_id, result) = process_input(in_file)
    archive_output(iteration, it_id, result)

def process_input(in_file):
    """
    Read the input data and process it to get the results.
    """
    (iteration, it_id, x, y, z, w) = read_input(in_file)

    # This is an example, we just calculate a 4D parabola...
    result = x**2 + y**2 + w**2 + z**2

    return (iteration, it_id, result)

def write_output(out_file, iteration, it_id, result):
    """
    Write the output file containing the calculated results.
    """
    # We write the XML "by-hand", as ElementTree does not support "pretty-printing"
    header = """\
    <!--
        Results file created by """ + command_name() + """
        on """ + current_time() + """
        for iteration """ + str(iteration) + """
    -->
    <batch>
      <dataType>run_results</dataType>
      <runID>0</runID>
      <nIndividuals>1</nIndividuals>
      <individuals>
        <individual0>
          <iteration>""" + str(iteration) + """</iteration>
          <id>""" + str(it_id) + """</id>
          <isValid>true</isValid>
          <isDirty>false</isDirty>
          <nResults>1</nResults>
          <results>
            <rawResult0>""" + str(result) + """</rawResult0>
          </results>
        </individual0>
      </individuals>
    </batch>
    """
    with open(out_file, 'w') as ofile:
        ofile.write(dedent(header))

def archive_output(iteration, it_id, result):
    """
    Archive the calculated results (for instance, this could mean
    to store the results in a DB).
    """
    print("Archiving the results...")
    # Nothing to do at the moment

def read_input(in_file):
    """
    Read the input file and extract the parameters for the run.
    """

    xmltree = None
    try:
        with open(in_file) as ifile:
            xmltree = ET.parse(ifile)
    except ET.ParseError as e:
        sys.exit("\nERROR: parsing error on input file '"
                 + in_file.name + "':\n\t" + str(e))

    root = xmltree.getroot()
    #
    # More error handling needed here, in case the XML file
    # doesn't have the expected structure
    #
    ind = root.find("./individuals/individual0")
    try:
        iteration = int(ind.find("./iteration").text)
    except:
        iteration = -1

    try:
        it_id = ind.find("./id").text
    except:
        it_id = "UNKNOWN_ID"

    # nr_results = int(root.find("./nResults").text)
    nr_params = int(ind.find("./nVars").text)

    # if nr_results != 1:
    #    sys.exit("\nERROR: unexpected parameter: nResults=" + str(nr_results) + " (expected 1)!")
    if nr_params != 4:
        sys.exit("\nERROR: unexpected parameter: nVars=" + str(nr_params) + " (expected 4)!")

    params = ind.findall(".//*[isLeaf]")
    if len(params) != nr_params:
        sys.exit("\nERROR: inconsistent data in input file: nVars=" + str(nr_params)
                 + ", but found " + str(len(params)) + " parameters!")

    x = float(ind.find("./vars/var0").find("values/value0").text)
    y = float(ind.find("./vars/var1").find("values/value0").text)
    z = float(ind.find("./vars/var2").find("values/value0").text)
    w = float(ind.find("./vars/var3").find("values/value0").text)

    return (iteration, it_id, x, y, z, w)


def main(argv):
    """
    This script encapsulates all the problem-specific information for an optimization
    problem. It is called by the optimizer to evaluate some parameter set, and provides
    the results in an output file.
    """

    parser = argparse.ArgumentParser(description="External evaluator")

    parser.add_argument("--initvalues", help="Start with the given initial values,"
                        " either 'min' or 'max' (default: 'random')")

    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--init", help="Perform initialization work",
                       action="store_true")
    group.add_argument("--finalize", help="Perform finalization work",
                       action="store_true")
    group.add_argument("--setup", help="Provide initialization data to the caller",
                       action="store_true")
    group.add_argument("--evaluate", help="Evaluate the parameters from the input file",
                       action="store_true")
    group.add_argument("--archive", help="Archive the results",
                       action="store_true")

    parser.add_argument("--data", help="Read the input data from the given file"
                        " (default: 'data.xml')",
                        metavar='IN_FILE', default="data.xml")
    parser.add_argument("--result", help="Write the results to the given file"
                        " (default: 'output.xml')",
                        metavar='OUT_FILE', default="output.xml")

    args = parser.parse_args()

    # These options checks are not supported by argparse (in Python 2.7 at least)
    if args.initvalues and not args.setup:
        sys.exit('The option --initvalues may only be used with --setup! Aborting...')

    initial_values = args.initvalues or "random"
    if initial_values not in set(["min", "max", "random"]):
        sys.exit("Wrong --initvalues value, use either 'min' or 'max'! Aborting...")

    # Now do our job
    if args.init:
        init()
    elif args.finalize:
        finalize()
    elif args.setup:
        setup(args.result, initial_values)
    elif args.evaluate:
        evaluate(args.data, args.result)
    elif args.archive:
        archive(args.data)
    else:
        sys.exit("ERROR: inconsistent options! why are we here?")


def command_name():
    """Return a cleaned up name of this program."""
    return os.path.basename(sys.argv[0])

def current_time():
    """Return the current date/time as a string."""
    return str(datetime.now())

if __name__ == "__main__":
    main(sys.argv[1:])
